// @pnpm/core@10.2.8/lib/link/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extendOptions = void 0;
const path_1 = __importDefault(require("path"));
const normalize_registries_1 = require("@pnpm/normalize-registries");
async function extendOptions(opts) {
    if (opts) {
        for (const key in opts) {
            if (opts[key] === undefined) {
                delete opts[key];
            }
        }
    }
    const defaultOpts = await defaults(opts);
    const extendedOpts = { ...defaultOpts, ...opts, storeDir: defaultOpts.storeDir };
    extendedOpts.registries = (0, normalize_registries_1.normalizeRegistries)(extendedOpts.registries);
    return extendedOpts;
}
exports.extendOptions = extendOptions;
async function defaults(opts) {
    const dir = opts.dir ?? process.cwd();
    return {
        binsDir: path_1.default.join(dir, 'node_modules', '.bin'),
        dir,
        force: false,
        forceSharedLockfile: false,
        hoistPattern: undefined,
        lockfileDir: opts.lockfileDir ?? dir,
        nodeLinker: 'isolated',
        registries: normalize_registries_1.DEFAULT_REGISTRIES,
        storeController: opts.storeController,
        storeDir: opts.storeDir,
        useLockfile: true,
    };
}
