// @pnpm/server@15.0.2/lib/connectStoreController.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.connectStoreController = void 0;
const fetch_1 = require("@pnpm/fetch");
const p_limit_1 = __importDefault(require("p-limit"));
const promise_share_1 = __importDefault(require("promise-share"));
const uuid_1 = require("uuid");
async function connectStoreController(initOpts) {
    const remotePrefix = initOpts.remotePrefix;
    const limitedFetch = limitFetch.bind(null, (0, p_limit_1.default)(initOpts.concurrency ?? 100));
    return new Promise((resolve, reject) => {
        resolve({
            close: async () => { },
            fetchPackage: fetchPackage.bind(null, remotePrefix, limitedFetch),
            getFilesIndexFilePath: () => ({ filesIndexFile: '', target: '' }),
            importPackage: async (to, opts) => {
                return limitedFetch(`${remotePrefix}/importPackage`, {
                    opts,
                    to,
                });
            },
            prune: async () => {
                await limitedFetch(`${remotePrefix}/prune`, {});
            },
            requestPackage: requestPackage.bind(null, remotePrefix, limitedFetch),
            stop: async () => {
                await limitedFetch(`${remotePrefix}/stop`, {});
            },
            upload: async (builtPkgLocation, opts) => {
                await limitedFetch(`${remotePrefix}/upload`, {
                    builtPkgLocation,
                    opts,
                });
            },
        });
    });
}
exports.connectStoreController = connectStoreController;
function limitFetch(limit, url, body) {
    return limit(async () => {
        // TODO: the http://unix: should be also supported by the fetcher
        // but it fails with node-fetch-unix as of v2.3.0
        if (url.startsWith('http://unix:')) {
            url = url.replace('http://unix:', 'unix:');
        }
        const response = await (0, fetch_1.fetch)(url, {
            body: JSON.stringify(body),
            headers: { 'Content-Type': 'application/json' },
            method: 'POST',
            retry: {
                retries: 100,
            },
        });
        if (!response.ok) {
            throw await response.json();
        }
        const json = await response.json(); // eslint-disable-line
        if (json.error) {
            throw json.error;
        }
        return json;
    });
}
async function requestPackage(remotePrefix, limitedFetch, // eslint-disable-line
wantedDependency, options) {
    const msgId = (0, uuid_1.v4)();
    return limitedFetch(`${remotePrefix}/requestPackage`, {
        msgId,
        options,
        wantedDependency,
    })
        .then((packageResponseBody) => {
        const fetchingBundledManifest = !packageResponseBody['fetchingBundledManifestInProgress'] // eslint-disable-line
            ? undefined
            : limitedFetch(`${remotePrefix}/rawManifestResponse`, {
                msgId,
            });
        delete packageResponseBody['fetchingBundledManifestInProgress']; // eslint-disable-line
        if (options.skipFetch) {
            return {
                body: packageResponseBody,
                bundledManifest: fetchingBundledManifest && (0, promise_share_1.default)(fetchingBundledManifest),
            };
        }
        const fetchingFiles = limitedFetch(`${remotePrefix}/packageFilesResponse`, {
            msgId,
        });
        return {
            body: packageResponseBody,
            bundledManifest: fetchingBundledManifest && (0, promise_share_1.default)(fetchingBundledManifest),
            files: (0, promise_share_1.default)(fetchingFiles),
            finishing: (0, promise_share_1.default)(Promise.all([fetchingBundledManifest, fetchingFiles]).then(() => undefined)),
        };
    });
}
function fetchPackage(remotePrefix, limitedFetch, // eslint-disable-line
options) {
    const msgId = (0, uuid_1.v4)();
    return limitedFetch(`${remotePrefix}/fetchPackage`, {
        msgId,
        options,
    })
        .then((fetchResponseBody) => {
        const fetchingBundledManifest = options.fetchRawManifest
            ? limitedFetch(`${remotePrefix}/rawManifestResponse`, { msgId })
            : undefined;
        const fetchingFiles = limitedFetch(`${remotePrefix}/packageFilesResponse`, {
            msgId,
        });
        return {
            bundledManifest: fetchingBundledManifest && (0, promise_share_1.default)(fetchingBundledManifest),
            files: (0, promise_share_1.default)(fetchingFiles),
            filesIndexFile: fetchResponseBody.filesIndexFile,
            finishing: (0, promise_share_1.default)(Promise.all([fetchingBundledManifest, fetchingFiles]).then(() => undefined)),
            inStoreLocation: fetchResponseBody.inStoreLocation,
        };
    });
}
